### DISCUSSION TITLE
I think that OnAfterBootstrap event should fire before Cobra command setup and execution

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1267] [Q&A]
@ganigeorgiev 

I'm not sure that this is a good idea in pocketbase.go.

```golang
	pb.RootCmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
		return pb.Bootstrap()
	}
```

Why can't the pocketbase bootstrap be done in pocketbase.go Execute?  Just put it before root command execute.

```golang
	go func() {
		defer wg.Done()
		if err := pb.RootCmd.Execute(); err != nil {
			log.Println(err)
		}
	}()
```

In particular, I'm finding it to be impossible to usefully intervene between beforeBootstrap and afterBootstrap.  As it is now, afterBootstrap is too late.  The issue is that I want to be able to create my own settings struct and `Env` struct for services that I'll need across all Cobra commands.  This struct (which might become a DI container of sorts) needs to be filled out sooner than I'm able to.  Due to Pocketbase using the persistent pre run event already, it's tricky to get in there somehow.

Basically, I don't like the granularity of pocketbase startup events.  I think that there are big flaws around "As a framework" at this time and that the root command and that the bootstrap events aren't right.

I want to point out that from my own struggles I can say that Pocketbase using a Cobra root command `PersistentPreRunE` function is hugely problematic for use as a framework because this hugely interferes with me being able to setup a common bootstrap across all custom commands.  I wasted a lot of time working through the oddities of how `PersistentPreRun` works when it's added to the top-most command (without a parent) and other sub root commands.  I mean that the work around logic for calling parents is too tricky.

As it is now, I think that the setup/init/bootstrap it's just wrong, and that I'm finding myself needing to work around too much.

**Comment by ganigeorgiev** [comment]
I'm sorry, I don't understand the described issue.

`OnBeforeBootstrap` and `OnAfterBootstrap` event hooks were added recently with v0.9 and they indicate the period right before and after app initialization, aka. after `OnAfterBootstrap` it is safe to access the database and the app settings.

We can change the behavior but I'm not sure what is the issue here. Could you give a concrete example what you are trying to do?

**Reply by dschissler** [reply]
Thanks for the super fast response.  The problem here is that `OnAfterBootstrap` is happening after the root command has already been executed.  This makes it impossible to access a loaded up `app` to able to wrap it up so that custom commands can be added.

**Reply by dschissler** [reply]
This is just currently what I'm working on.  It's a global app config.  It will change as I integrate Viper and default configs from embed, but this is how it is now.

I'm not able to get something as simple as the app name `settings.Meta.AppName` for the app config that is then passed into each of my custom commands.  Pocketbase current requires that I setup the custom commands before `app` has been bootstraped.


```golang
type Config struct {
	DataDir  string
	Settings *settings.Settings
	SSH      SSH
	Web      Web
}

func New(app core.App) *Config {
	dataDir := app.DataDir()
	settings := app.Settings()

	// This is too early, if done differently then I can't use it for setting up custom Cobra commands
	appName := strings.ToLower(settings.Meta.AppName)

	return &Config{
		Settings: settings,
		DataDir:  dataDir,
		Web: Web{
			PIDPath: "/tmp/" + appName + "_web.pid",
		},
		SSH: SSH{
			Dir:                   dataDir + "/ssh",
			KeyPrefix:             "id_ed25519",
			KnownClientsFileName:  "clients_known",
			KnownInstancesDirName: "instances_known",
			Port:                  2222,
			PIDPath:               "/tmp/" + appName + "_ssh.pid",
		},
	}
}
```

**Reply by ganigeorgiev** [reply]
The event hooks are not tied to the cobra instance since they execute on the `core.App` interface.

You can register custom commands by directly calling `app.RootCmd.AddCommand(myCommand(app))`.

**Reply by ganigeorgiev** [reply]
@dschissler I still don't understanding what you want to achieve exactly. Suggestions for improvements are welcomed. Feel free to open PR and I'll review it.

**Reply by dschissler** [reply]
> You can register custom commands by directly calling app.RootCmd.AddCommand(myCommand(app)).

The problem is that I don't want to pass in `app` to a command.

```golang
func main() {
	app := pocketbase.New()

	cmd.Init(app)

	if err := app.Execute(); err != nil {
		log.Fatal(err)
	}
}
```

You can see that `env` takes `app` to be setup.

```golang
func Init(app *pocketbase.PocketBase) {
	app.OnAfterBootstrap().Add(func(e *core.BootstrapEvent) error {
		env := env.New(app)
		initModelEvents(env)

		rootCmd := app.RootCmd
		rootCmd.AddCommand(NewWebserverCommand(env))
		rootCmd.AddCommand(NewInstanceCommand(env))
		rootCmd.AddCommand(NewInstallCommand(env))

		rootCmd.AddCommand(pb_cmd.NewTempUpgradeCommand(app))

		return nil
	})

}
```

**Reply by ganigeorgiev** [reply]
@dschissler We can try to move pb.Bootstrap() out of the cobra PreRun and execute it right after `pocketbase.New()` (I don't remember exactly why I've done it this way; maybe it was before the eager flag parsing was added). Is this what you want?

**Reply by dschissler** [reply]
@ganigeorgiev

> We can try to move pb.Bootstrap() out of the cobra PreRun and execute it right after pocketbase.New() ... Is this what you want?

I believe so.  Yes.

Will you move that code, or should I make a PR?  I'm not sure how to use git for Golang modules, but I'll try to figure it out.

**Reply by ganigeorgiev** [reply]
I remember now why it is attached to the PreRun - it is because of https://github.com/pocketbase/pocketbase/issues/404.

We can change it and initialize it directly on `pocketbase.New()` and eventually skip when manually checking the `os.Args` to determine whether `--version` or the `--help` flags were called.

I'll have to think a little more on this.

**Reply by dschissler** [reply]
Unless there are some edge cases, I think that this will open up two ways to access a bootstrapped app . The developer could add a custom persistent pre run (global pre run) to the root command (without having to fight against an existing one), and also just using the `OnAfterBootstrap` event.  Both of these have different use cases / timings.

**Reply by dschissler** [reply]
I'm going to look through the code to see what can be done.  It's kind of a blocker on me right now.

**Reply by ganigeorgiev** [reply]
As a temporary workaround, you can just manually call `app.Bootstrap()` after `pocketbase.New()`.

**Reply by dschissler** [reply]
Thanks.  As a work around I'll unset `rootCmd.PersistentPreRunE` and then call `app.Bootstrap()` manually.

**Reply by ganigeorgiev** [reply]
@dschissler

After some consideration, I'm not sure what would be the best approach here. The current behavior is correct and it is inline with the `New()` godoc:

```
// New creates a new PocketBase instance with the default configuration.
// Use [NewWithConfig()] if you want to provide a custom configuration.
//
// Note that the application will not be initialized/bootstrapped yet,
// aka. DB connections, migrations, app settings, etc. will not be accessible.
// Everything will be initialized when [Start()] is executed.
// If you want to initialize the application before calling [Start()],
// then you'll have to manually call [Bootstrap()].
func New() *PocketBase
```

Calling `Bootstrap()` inside the app factory practically will never allow attaching to the before and after bootstrap events since they will be triggered before the method return the app instance to the user.

I think the only thing we can optimize is to prevent calling Bootstrap() in the prerun if the app is already initialized (although it should still work even if you call Bootstrap() multiple times because it first resets the previous state before loading the new one).

**Reply by ganigeorgiev** [reply]
Alternatively, probably the better approach would be to revert the #404 change and attach directly the `Bootstrap` to the beginning of `pb.Start()/pb.Execute()` (with some additional manual checks for the `--version` and `--help` flags).

This way it will be triggered before all cobra callbacks and still allow users to attach to the before/after event hooks.

**Reply by dschissler** [reply]
I deleted my last post, and another one that I was writing because I realized it was the wrong path.

**Reply by dschissler** [reply]
@ganigeorgiev Does Cobra command have a way to test what the route will be?

**Reply by dschissler** [reply]
What about `Find`:

* [doc](https://pkg.go.dev/github.com/spf13/cobra?utm_source=godoc#Command.Find)
* [source](https://github.com/spf13/cobra/blob/v1.6.1/command.go#L677)

**Reply by ganigeorgiev** [reply]
Yes, I agree, calling inside `app.Execute()` would be better. I'll try to update the initialization process in the v0.10 release (_most likely next week since I'm having some difficulties with the multi-match filter queries_).

**Reply by ganigeorgiev** [reply]
> Does Cobra command have a way to test what the route will be?

I'm not sure but probably there are several ways we can determine the command/flags and ignore it. Placing it inside the PreRun was just the easiest way to fix #404.

**Reply by dschissler** [reply]
Ok, so I'm good now with the work around that you gave.

I looked through the [Cobra Find code](https://github.com/spf13/cobra/blob/v1.6.1/command.go#L677), and I think that you can call to this method to see if a returned command is named "help" (to not bootstrap and thus not creating pb_data), or by comparing this struct by reference, or whatever.  Call `RootCmd.Find` before `RootCmd.Execute`.  From what I can tell, the only downside is the overhead of command line routing twice.  Seems minimal.

So I'm think that here is the **critical** part.  That if a developer wants to add commands or to mess with the root command that this should be done in the before or after bootstrap events.  I mean, it seems so important that perhaps `RootCmd` should be made private as `rootCmd` and only exposed through these two (or more) events.  Bootstrap before would occur before bootstrap.  You'd be able to add commands to root command, but without a setup app.  Maybe that is only liability and no value, but I can't foresee every possibility there.  Although, maybe before bootstrap is only for configuring app settings, and not root command yet.  So, possibly after bootstrap could be the only place for root command to be exposed via event parameter.  At this point the app struct can be wrapped up so that a developer can make their own system (as they do) and to pass this enhanced environment with custom configs, DI lazy load services, etc into the command add function.

I'm thinking that making root command as private might just save you a great many pointless Discussion points in the future.  Well, at least you would just be able to point them to the docs and a foolproof way to do it.  Otherwise it's just riddled with gotchas.

**Reply by dschissler** [reply]
Also I want to point out that if you were to take my advice to make `RootCmd` private as `rootCmd`, and to only expose the root command on the afterBootstrap event, and to only fire afterBootstrap when a boostrap had occurred (obvious), then calling `rootCmd.Find` would be incredibly quick because the custom commands would not have been added yet.  It seems elegant to me.

So in my case, if `help` was being called, then none of the commands would even be added to the `rootCmd`, because the logic around `rootCmd.Find` would have determined not to bootstrap, and therefore, not to call the after bootstrap event where the commands are added.

**Reply by ganigeorgiev** [reply]
>  That if a developer wants to add commands or to mess with the root command that this should be done in the before or after bootstrap events. I mean, it seems so important that perhaps RootCmd should be made private as rootCmd and only exposed through these two (or more) events. 

No. `core.App` can exist without cobra. It is easier to test this way since you need only to load the server. There is nothing wrong if developers want to register a new command before initialization (that actually how we register the `serve` command). On command `Run` the app will be already initialized. If the user want access to the `app` instance they can pass it as an argument to the command wrapper function.

> maybe before bootstrap is only for configuring app settings, and not root command yet.

Yes, it was added in v0.9 to provide options to load custom initialization or doing something else before opening the database and loading the app settings (mostly for future "plugins"). It doesn't depend on the command line execution.

**Reply by dschissler** [reply]
That's fine.

**Reply by ganigeorgiev** [reply]
@dschissler _Just to let you know that I've pushed an update in `master`(https://github.com/pocketbase/pocketbase/commit/c25e67e13d345f94411e4ab137ec6feec93d9b40) that calls `app.Bootstrap()` before the cobra command exec flow. I was planning to ship the multi-match filter enhancement with v0.10, but I'll need a little more time to test it properly, so most likely I'll cut a v0.10 release with the applied changes sometime later this week._

**Reply by dschissler** [reply]
The application startup is much easier to read now.

**Comment by dschissler** [comment]
@ganigeorgiev I'm looking at the pocketbase.go file now.  I can see a pretty painless way to get Cobra out of this struct.  The largest hurdle was the `PersistentPreRunE` function.  Just to be clear, I know that you aren't sold on this being a good idea.

One day after 0.10 I'll play with breaking the root command out in an experimental branch.  At least you'll be able to see what that would look like.  Basically a bound `RootCmd.Execute` function would be passed into `pb.Execute`, and the Cobra setup could be called just as easily as it is now, but with either a longer module name or a longer `New` method.  Or also, it could be kept as it is now with just `New` and then changing the de-Cobraed pocketbase app to `NewApp`.  The easy case would, docs and everything wouldn't need to change.

The benefit would be that someone could embed Pocketbase deep within their app as just another mode / server, or they could use the competing CLI framework.  Probably the community would just do that somewhere.  At this point I'm just sticking with Cobra after switching from `urfave/cli` because the damage has already been done to my time.

**Comment by dschissler** [comment]
@ganigeorgiev Not that I'm losing any sleep over it, but because I'm adding commands so late, I now have a ~broken top level help listing~ completely broken help at all levels of my app.  This is a pattern/dev issue and not a Pocketbase one though.  One day I'll see if I can rearrange things to make it work properly.  It's more a curiosity than anything though.
